File: Leetcode 3801-3900/3855/3855.java
// Leetcode 3855: Sum of K-Digit Numbers in a Range
// https://leetcode.com/problems/sum-of-k-digit-numbers-in-a-range/
// Solved on 1st of March, 2026
class Solution {
    /**
     * Calculates the sum of all k-digit numbers formed within a given range [l, r].
     *
     * @param l The lower bound of the range.
     * @param r The upper bound of the range.
     * @param k The number of digits for the numbers to be summed.
     * @return The total sum modulo 10^9 + 7.
     */
    public int sumOfNumbers(int l, int r, int k) {
        long mod = 1000000007;
        long numDigits = r - l + 1;
        long sumOfDigits = (long) (l + r) * numDigits / 2;
        long termOne = sumOfDigits % mod;
        long termTwo = power(numDigits, k - 1, mod);
        long termThree = (power(10, k, mod) - 1 + mod) % mod;
        long inverseNine = power(9, mod - 2, mod);
        long result = (termOne * termTwo) % mod;
        result = (result * termThree) % mod;
        result = (result * inverseNine) % mod;
        return (int) result;
    }

    private long power(long base, long exp, long mod) {
        long res = 1;
        base = base % mod;
        while (exp > 0) {
            if ((exp & 1) == 1) {
                res = (res * base) % mod;
            }
            base = (base * base) % mod;
            exp >>= 1;
        }
        return res;
    }
}

File: Leetcode 3801-3900/3855/3855.py
# Leetcode 3855: Sum of K-Digit Numbers in a Range
# https://leetcode.com/problems/sum-of-k-digit-numbers-in-a-range/
# Solved on 1st of March, 2026
class Solution:
    def sumOfNumbers(self, l: int, r: int, k: int) -> int:
        """
        Calculates the sum of all k-digit numbers where each digit is in the range [l, r].

        :param l: The lower bound of the allowed digits (inclusive).
        :param r: The upper bound of the allowed digits (inclusive).
        :param k: The number of digits in each number.
        :return: The total sum of all valid k-digit numbers modulo 10^9 + 7.
        """
        modVal = 1000000007

        numChoices = r - l + 1
        sumDigits = (l + r) * numChoices // 2
        invNine = pow(9, modVal - 2, modVal)
        powTen = pow(10, k, modVal)
        repOnes = (powTen - 1) * invNine % modVal
        powChoices = pow(numChoices, k - 1, modVal)
        totalSum = (sumDigits * powChoices % modVal) * repOnes % modVal

        return totalSum